software/base/src/main/java/org/apache/brooklyn/entity/software/base/AbstractSoftwareProcessDriver.java

```diff
@@ -312,16 +312,36 @@ public void run() {
         } else {
             DynamicTasks.queue("pre-launch-command", new Runnable() { public void run() {
                 ServiceStateLogic.setExpectedState(getEntity(), Lifecycle.STARTING);
-                runPreLaunchCommand();
+                try {
+                    runPreLaunchCommand();
+                } catch (Exception e) {
+                    ServiceStateLogic.setExpectedState(getEntity(), Lifecycle.ON_FIRE);
+                    throw Exceptions.propagate(e);
+                }
             }});
             DynamicTasks.queue("launch (main)", new Runnable() { public void run() {
-                launch();
+                try {
+                    launch();
+                } catch (Exception e) {
+                    ServiceStateLogic.setExpectedState(getEntity(), Lifecycle.ON_FIRE);
+                    throw Exceptions.propagate(e);
+                }
             }});
             DynamicTasks.queue("post-launch-command", new Runnable() { public void run() {
-                runPostLaunchCommand();
+                try {
+                    runPostLaunchCommand();
+                } catch (Exception e) {
+                    ServiceStateLogic.setExpectedState(getEntity(), Lifecycle.ON_FIRE);
+                    throw Exceptions.propagate(e);
+                }
             }});
             DynamicTasks.queue("post-launch", new Runnable() { public void run() {
-                postLaunch();
+                try {
+                    postLaunch();
+                } catch (Exception e) {
+                    ServiceStateLogic.setExpectedState(getEntity(), Lifecycle.ON_FIRE);
+                    throw Exceptions.propagate(e);
+                }
```

Comment:
Still need to do the same for the `doFullStartOnRestart()` case. Suggest instead wrapping the `restart` effector as done in [`start`](https://github.com/apache/brooklyn-server/blob/master/software/base/src/main/java/org/apache/brooklyn/entity/software/base/lifecycle/MachineLifecycleEffectorTasks.java#L353-L363).